Nodes/HMC6343Node.cpp

```diff
@@ -0,0 +1,205 @@
+/****************************************************************************************
+ *
+ * File:
+ * 		HMC6343Node.cp
+ *
+ * Purpose:
+ *		The HMC6343 node communicates with a HMC6343 compass and provides CompassDataMsgs
+ *		to the message bus.
+ *
+ * Developer Notes:
+ *
+ *
+ ***************************************************************************************/
+
+#include "HMC6343Node.h"
+#include "Messages/CompassDataMsg.h"
+#include "logger/Logger.h"
+#include "wiringPi.h"
+#include "utility/Utility.h"
+
+// For std::this_thread
+#include <chrono>
+#include <thread>
+
+
+// The device datasheet mentions that this is 0x32. In our case this doesn't seem to be
+// correct, so stick with 0x19.
+#define I2C_ADDRESS 			0x19
+#define I2C_DATASHEET_ADDRESS	0x32
+
+// HMC6343 Commands
+#define COM_POST_HEADING 		0x50
+#define COM_POST_TILT 			0x55
+#define COM_POST_MAG 			0x45
+#define COM_POST_ACCEL 			0x40
+#define COM_READ_EEPROM 		0xE1
+
+#define EEPROM_ADDRESS			0x00
+
+
+#define COM_ORIENT_LEVEL 0x72
+#define COM_ORIENT_SIDEWAYS 0x73
+#define COM_ORIENT_FLATFRONT 0x74
+
+// HMC6343 Orientations
+#define LEVEL 0 // X = forward, +Z = up (default)
+#define SIDEWAYS 1 // X = forward, +Y = up
+#define FLATFRONT 2 // Z = forward, -X = up
+
+
+#define COMPASS_SENSOR_SLEEP_MS	100
+
+
+HMC6343Node::HMC6343Node(MessageBus& msgBus, const int headingBufferSize)
+: ActiveNode(NodeID::Compass, msgBus), m_I2C(), m_Initialised(false), m_HeadingBufferSize(headingBufferSize)
+{
+
+}
+
+
+bool HMC6343Node::init()
+{
+	m_Initialised = false;
+
+	if(m_I2C.init(I2C_ADDRESS))
+	{
+		m_I2C.beginTransmission();
+		m_I2C.writeReg(COM_READ_EEPROM, EEPROM_ADDRESS);
+		delay(10);
+		uint8_t deviceID = m_I2C.read();
+
+		m_I2C.endTransmission();
+
+		// The Device reports the I2C address that is mentioned in the datasheet
+		if(deviceID == I2C_DATASHEET_ADDRESS)
+		{
+			m_Initialised = true;
+		}
+		else
+		{
+			Logger::error("%s Failed to successfully read from the HMC6343 compass, check connection!", __PRETTY_FUNCTION__);
+		}
+	}
+	else
+	{
+		Logger::error("%s Failed to obtain I2C file descriptor", __PRETTY_FUNCTION__);
+	}
+
+	return m_Initialised;
+}
+
+
+void HMC6343Node::start()
+{
+	if(m_Initialised)
+	{
+		runThread(HMC6343ThreadFunc);
+	}
+	else
+	{
+		Logger::error("%s Cannot start compass sensor thread as the node was not correctly initialised!", __PRETTY_FUNCTION__);
+	}
+}
+
+void HMC6343Node::processMessage(const Message* msg)
+{ }
+
+
+bool HMC6343Node::readData(float& heading, float& pitch, float& roll)
+{
+	const int BYTES_TO_READ = 6;
+	uint8_t buffer[BYTES_TO_READ];
+
+	if(m_Initialised)
+	{
+		// Extract the data from the compass, each piece of data is made up of two bytes, and their are
+		// 3 pieces of data.
+		m_I2C.beginTransmission();
+		m_I2C.writeReg(COM_POST_HEADING, 0x00);
```

Comment:
why use writeReg here because to get the value you don't need to write on registry see [old code] (https://github.com/AlandSailingRobots/sailingrobot/blob/master/Compass/HMC6343.cpp#L124).